# This notebook contains all parts (A to D) of the assignment

## The final results and discussion of them is located in the end of the notebook

# A. Build a baseline model (5 marks)

###  Download and Clean Dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.shape

(1030, 9)

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

### Spliting data into predictors and target

In [5]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [6]:
n_cols = predictors.shape[1] # number of predictors

### Importing Keras

In [7]:
import keras

Using TensorFlow backend.


In [8]:
from keras.models import Sequential
from keras.layers import Dense

### Building Neural Network

In [9]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

model_A = regression_model()

Instructions for updating:
Colocations handled automatically by placer.


## 1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3)

y_test.head()

649     6.90
261    37.40
691     6.81
665    11.39
397    29.16
Name: Strength, dtype: float64

## 2. Train the model on the training data using 50 epochs.

In [11]:
model_A.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=2)

Instructions for updating:
Use tf.cast instead.
Train on 504 samples, validate on 217 samples
Epoch 1/50
 - 1s - loss: 70153.2899 - val_loss: 59566.3423
Epoch 2/50
 - 0s - loss: 45656.2183 - val_loss: 36326.3362
Epoch 3/50
 - 0s - loss: 24357.8762 - val_loss: 17141.8294
Epoch 4/50
 - 0s - loss: 9877.2698 - val_loss: 6525.7569
Epoch 5/50
 - 0s - loss: 3928.2232 - val_loss: 3373.6609
Epoch 6/50
 - 0s - loss: 2659.7603 - val_loss: 2798.3694
Epoch 7/50
 - 0s - loss: 2407.9711 - val_loss: 2547.5404
Epoch 8/50
 - 0s - loss: 2163.6121 - val_loss: 2336.5057
Epoch 9/50
 - 0s - loss: 1953.6541 - val_loss: 2151.1663
Epoch 10/50
 - 0s - loss: 1772.1013 - val_loss: 1961.7454
Epoch 11/50
 - 0s - loss: 1603.8078 - val_loss: 1791.6043
Epoch 12/50
 - 0s - loss: 1456.3085 - val_loss: 1633.3479
Epoch 13/50
 - 0s - loss: 1324.8309 - val_loss: 1495.5580
Epoch 14/50
 - 0s - loss: 1208.9360 - val_loss: 1359.1737
Epoch 15/50
 - 0s - loss: 1105.2889 - val_loss: 1253.9058
Epoch 16/50
 - 0s - loss: 1014.8756 - v

## 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [12]:
from sklearn.metrics import mean_squared_error

y_pred = model_A.predict(X_test)

mean_squared_error(y_test, y_pred)

254.36556523173329

## 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [13]:
model_A = regression_model()

mse_list = []

def get_mse(model):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3)
    model.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    return mean_squared_error(y_test, y_pred)

In [14]:
for i in range(50):
    print('Computing {}th iteration'.format(i+1))
    mse_list.append(get_mse(model_A))

mse_list[:10]

Computing 1th iteration
Computing 2th iteration
Computing 3th iteration
Computing 4th iteration
Computing 5th iteration
Computing 6th iteration
Computing 7th iteration
Computing 8th iteration
Computing 9th iteration
Computing 10th iteration
Computing 11th iteration
Computing 12th iteration
Computing 13th iteration
Computing 14th iteration
Computing 15th iteration
Computing 16th iteration
Computing 17th iteration
Computing 18th iteration
Computing 19th iteration
Computing 20th iteration
Computing 21th iteration
Computing 22th iteration
Computing 23th iteration
Computing 24th iteration
Computing 25th iteration
Computing 26th iteration
Computing 27th iteration
Computing 28th iteration
Computing 29th iteration
Computing 30th iteration
Computing 31th iteration
Computing 32th iteration
Computing 33th iteration
Computing 34th iteration
Computing 35th iteration
Computing 36th iteration
Computing 37th iteration
Computing 38th iteration
Computing 39th iteration
Computing 40th iteration
Computing

[1655.0776659975893,
 718.1644832554562,
 346.38199079324306,
 252.22034675706377,
 172.16386256060673,
 148.98770160633978,
 127.72497241964028,
 109.40918417690312,
 111.79102723477074,
 107.4755316681148]

## 5. Report the mean and the standard deviation of the mean squared errors.

In [15]:
mean_mse = np.mean(mse_list)
print('Mean MSE: ', mean_mse)

Mean MSE:  119.68541321852037


In [16]:
sd_mse = np.std(mse_list)
print('Standard deviation of MSE: ', sd_mse)

Standard deviation of MSE:  243.11265705747866


# B. Normalize the data (5 marks)

In [17]:
model_B = regression_model()
predictors_norm = (predictors - predictors.mean()) / predictors.std()

In [18]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3)

y_test.head()

104    33.40
181    82.60
432    26.20
327    55.83
667    39.32
Name: Strength, dtype: float64

In [19]:
model_B.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=2)

Train on 504 samples, validate on 217 samples
Epoch 1/50
 - 1s - loss: 1460.1302 - val_loss: 1630.0080
Epoch 2/50
 - 0s - loss: 1446.7574 - val_loss: 1614.6188
Epoch 3/50
 - 0s - loss: 1433.5008 - val_loss: 1598.7402
Epoch 4/50
 - 0s - loss: 1419.7812 - val_loss: 1582.8624
Epoch 5/50
 - 0s - loss: 1406.1211 - val_loss: 1566.5995
Epoch 6/50
 - 0s - loss: 1392.0765 - val_loss: 1550.1299
Epoch 7/50
 - 0s - loss: 1377.9561 - val_loss: 1533.2889
Epoch 8/50
 - 0s - loss: 1363.5533 - val_loss: 1516.2445
Epoch 9/50
 - 0s - loss: 1349.0681 - val_loss: 1498.8921
Epoch 10/50
 - 0s - loss: 1334.2527 - val_loss: 1480.9766
Epoch 11/50
 - 0s - loss: 1319.1057 - val_loss: 1462.8503
Epoch 12/50
 - 0s - loss: 1303.7499 - val_loss: 1444.5214
Epoch 13/50
 - 0s - loss: 1288.0521 - val_loss: 1425.8150
Epoch 14/50
 - 0s - loss: 1272.1434 - val_loss: 1406.6305
Epoch 15/50
 - 0s - loss: 1255.8772 - val_loss: 1387.3665
Epoch 16/50
 - 0s - loss: 1239.6458 - val_loss: 1367.2684
Epoch 17/50
 - 0s - loss: 1222.7957

In [20]:
y_pred = model_B.predict(X_test)

mean_squared_error(y_test, y_pred)

608.6309167445229

In [21]:
model_B = regression_model()

mse_list_norm = []

for i in range(50):
    print('Computing {}th iteration'.format(i+1))
    mse_list_norm.append(get_mse(model_B))

mse_list_norm[:10]

Computing 1th iteration
Computing 2th iteration
Computing 3th iteration
Computing 4th iteration
Computing 5th iteration
Computing 6th iteration
Computing 7th iteration
Computing 8th iteration
Computing 9th iteration
Computing 10th iteration
Computing 11th iteration
Computing 12th iteration
Computing 13th iteration
Computing 14th iteration
Computing 15th iteration
Computing 16th iteration
Computing 17th iteration
Computing 18th iteration
Computing 19th iteration
Computing 20th iteration
Computing 21th iteration
Computing 22th iteration
Computing 23th iteration
Computing 24th iteration
Computing 25th iteration
Computing 26th iteration
Computing 27th iteration
Computing 28th iteration
Computing 29th iteration
Computing 30th iteration
Computing 31th iteration
Computing 32th iteration
Computing 33th iteration
Computing 34th iteration
Computing 35th iteration
Computing 36th iteration
Computing 37th iteration
Computing 38th iteration
Computing 39th iteration
Computing 40th iteration
Computing

[1940.7836348626545,
 426.97499808884015,
 148.31205822476576,
 115.4163074421897,
 108.65996053999318,
 99.6507609369059,
 113.00730630876568,
 105.47196514144396,
 135.76659651518776,
 108.76861659451683]

In [22]:
mean_mse_norm = np.mean(mse_list_norm)
print('Mean MSE (normalized data): ', mean_mse_norm)

Mean MSE (normalized data):  157.21599365438797


In [23]:
sd_mse_norm = np.std(mse_list_norm)
print('Standard deviation of MSE (normalized data): ', sd_mse_norm)

Standard deviation of MSE (normalized data):  258.8183682766227


# C. Increate the number of epochs (5 marks)

In [24]:
model_C = regression_model()

In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3)

y_test.head()

904    23.79
12     43.01
205    20.92
726    14.31
921     9.74
Name: Strength, dtype: float64

In [26]:
model_C.fit(X_train, y_train, validation_split=0.3, epochs=100, verbose=2)

Train on 504 samples, validate on 217 samples
Epoch 1/100
 - 1s - loss: 1610.3513 - val_loss: 1613.2737
Epoch 2/100
 - 0s - loss: 1599.9624 - val_loss: 1603.1090
Epoch 3/100
 - 0s - loss: 1589.4542 - val_loss: 1592.9906
Epoch 4/100
 - 0s - loss: 1579.0871 - val_loss: 1582.8861
Epoch 5/100
 - 0s - loss: 1568.5371 - val_loss: 1572.7612
Epoch 6/100
 - 0s - loss: 1557.7973 - val_loss: 1562.3704
Epoch 7/100
 - 0s - loss: 1546.8532 - val_loss: 1551.7791
Epoch 8/100
 - 0s - loss: 1535.6992 - val_loss: 1540.7741
Epoch 9/100
 - 0s - loss: 1524.0999 - val_loss: 1529.5051
Epoch 10/100
 - 0s - loss: 1512.0179 - val_loss: 1517.8222
Epoch 11/100
 - 0s - loss: 1499.5982 - val_loss: 1505.5677
Epoch 12/100
 - 0s - loss: 1486.5896 - val_loss: 1492.8444
Epoch 13/100
 - 0s - loss: 1473.0103 - val_loss: 1479.6026
Epoch 14/100
 - 0s - loss: 1458.6919 - val_loss: 1466.0482
Epoch 15/100
 - 0s - loss: 1443.8732 - val_loss: 1451.6922
Epoch 16/100
 - 0s - loss: 1428.3541 - val_loss: 1436.6146
Epoch 17/100
 - 0s 

In [27]:
y_pred = model_C.predict(X_test)

mean_squared_error(y_test, y_pred)

199.8313352050866

In [28]:
model_C = regression_model()

mse_list_norm_100 = []

for i in range(100):
    print('Computing {}th iteration'.format(i+1))
    mse_list_norm_100.append(get_mse(model_C))

mse_list_norm_100[:10]

Computing 1th iteration
Computing 2th iteration
Computing 3th iteration
Computing 4th iteration
Computing 5th iteration
Computing 6th iteration
Computing 7th iteration
Computing 8th iteration
Computing 9th iteration
Computing 10th iteration
Computing 11th iteration
Computing 12th iteration
Computing 13th iteration
Computing 14th iteration
Computing 15th iteration
Computing 16th iteration
Computing 17th iteration
Computing 18th iteration
Computing 19th iteration
Computing 20th iteration
Computing 21th iteration
Computing 22th iteration
Computing 23th iteration
Computing 24th iteration
Computing 25th iteration
Computing 26th iteration
Computing 27th iteration
Computing 28th iteration
Computing 29th iteration
Computing 30th iteration
Computing 31th iteration
Computing 32th iteration
Computing 33th iteration
Computing 34th iteration
Computing 35th iteration
Computing 36th iteration
Computing 37th iteration
Computing 38th iteration
Computing 39th iteration
Computing 40th iteration
Computing

[428.14772717679233,
 224.1286517443647,
 191.95574014393026,
 129.25070220818273,
 129.97755624494087,
 130.59632267333876,
 94.32154268560322,
 95.40381483930102,
 88.1891039727837,
 79.2078707072155]

In [29]:
mean_mse_norm_100 = np.mean(mse_list_norm_100)
print('Mean MSE (normalized data, 100 epochs): ', mean_mse_norm_100)

Mean MSE (normalized data, 100 epochs):  63.406545922921715


In [30]:
sd_mse_norm_100 = np.std(mse_list_norm_100)
print('Standard deviation of MSE (normalized data, 100 epochs): ', sd_mse_norm_100)

Standard deviation of MSE (normalized data, 100 epochs):  45.327161124465874


# D. Increase the number of hidden layers (5 marks)

In [31]:
def regression_model_D():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

model_D = regression_model_D()

In [32]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3)

y_test.head()

952    33.72
320    28.68
283    44.21
93     46.20
649     6.90
Name: Strength, dtype: float64

In [33]:
model_D.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=2)

Train on 504 samples, validate on 217 samples
Epoch 1/50
 - 1s - loss: 1529.8105 - val_loss: 1614.0066
Epoch 2/50
 - 0s - loss: 1516.0611 - val_loss: 1596.2363
Epoch 3/50
 - 0s - loss: 1497.5163 - val_loss: 1571.5700
Epoch 4/50
 - 0s - loss: 1472.1143 - val_loss: 1537.9184
Epoch 5/50
 - 0s - loss: 1436.6182 - val_loss: 1492.4006
Epoch 6/50
 - 0s - loss: 1389.5325 - val_loss: 1430.6044
Epoch 7/50
 - 0s - loss: 1326.9590 - val_loss: 1350.1988
Epoch 8/50
 - 0s - loss: 1246.1138 - val_loss: 1247.1167
Epoch 9/50
 - 0s - loss: 1142.6503 - val_loss: 1122.0313
Epoch 10/50
 - 0s - loss: 1017.8396 - val_loss: 972.1095
Epoch 11/50
 - 0s - loss: 871.1309 - val_loss: 808.5724
Epoch 12/50
 - 0s - loss: 714.8836 - val_loss: 648.6611
Epoch 13/50
 - 0s - loss: 568.8606 - val_loss: 509.6644
Epoch 14/50
 - 0s - loss: 445.0355 - val_loss: 412.0819
Epoch 15/50
 - 0s - loss: 362.6222 - val_loss: 353.5368
Epoch 16/50
 - 0s - loss: 303.9626 - val_loss: 322.7721
Epoch 17/50
 - 0s - loss: 271.0425 - val_loss: 3

In [34]:
y_pred = model_D.predict(X_test)

mean_squared_error(y_test, y_pred)

166.73091505089397

In [35]:
model_D = regression_model_D()

mse_list_norm_D = []

for i in range(50):
    print('Computing {}th iteration'.format(i+1))
    mse_list_norm_D.append(get_mse(model_D))

mse_list_norm_D[:10]

Computing 1th iteration
Computing 2th iteration
Computing 3th iteration
Computing 4th iteration
Computing 5th iteration
Computing 6th iteration
Computing 7th iteration
Computing 8th iteration
Computing 9th iteration
Computing 10th iteration
Computing 11th iteration
Computing 12th iteration
Computing 13th iteration
Computing 14th iteration
Computing 15th iteration
Computing 16th iteration
Computing 17th iteration
Computing 18th iteration
Computing 19th iteration
Computing 20th iteration
Computing 21th iteration
Computing 22th iteration
Computing 23th iteration
Computing 24th iteration
Computing 25th iteration
Computing 26th iteration
Computing 27th iteration
Computing 28th iteration
Computing 29th iteration
Computing 30th iteration
Computing 31th iteration
Computing 32th iteration
Computing 33th iteration
Computing 34th iteration
Computing 35th iteration
Computing 36th iteration
Computing 37th iteration
Computing 38th iteration
Computing 39th iteration
Computing 40th iteration
Computing

[265.5886684249818,
 202.09580118404935,
 134.62819338838645,
 116.70271350457213,
 126.76637476857358,
 107.71667847820542,
 109.13046289790641,
 111.23439924496617,
 109.27883191416988,
 114.79382402027636]

In [36]:
mean_mse_norm_D = np.mean(mse_list_norm_D)
print('Mean MSE (normalized data, 3 hidden layers): ', mean_mse_norm_D)

Mean MSE (normalized data, 3 hidden layers):  86.38566381434345


In [37]:
sd_mse_norm_D = np.std(mse_list_norm_D)
print('Standard deviation of MSE (normalized data, 3 hidden layers): ', sd_mse_norm_D)

Standard deviation of MSE (normalized data, 3 hidden layers):  46.08313789222228


# Stats

In [38]:
print('\nA')
print('Mean MSE: ', mean_mse)
print('Standard deviation of MSE: ', sd_mse)

print('\nB')
print('Mean MSE (normalized data): ', mean_mse_norm)
print('Standard deviation of MSE (normalized data): ', sd_mse_norm)

print('\nC')
print('Mean MSE (normalized data, 100 epochs): ', mean_mse_norm_100)
print('Standard deviation of MSE (normalized data, 100 epochs): ', sd_mse_norm_100)

print('\nD')
print('Mean MSE (normalized data, 3 hidden layers): ', mean_mse_norm_D)
print('Standard deviation of MSE (normalized data, 3 hidden layers): ', sd_mse_norm_D)


A
Mean MSE:  119.68541321852037
Standard deviation of MSE:  243.11265705747866

B
Mean MSE (normalized data):  157.21599365438797
Standard deviation of MSE (normalized data):  258.8183682766227

C
Mean MSE (normalized data, 100 epochs):  63.406545922921715
Standard deviation of MSE (normalized data, 100 epochs):  45.327161124465874

D
Mean MSE (normalized data, 3 hidden layers):  86.38566381434345
Standard deviation of MSE (normalized data, 3 hidden layers):  46.08313789222228


### Normalizing the data seemed to lower the accuracy of the model, as seen from comparison of Mean and Standard deviation of MSE from parts A and B

### Doubling the number of epochs had a significant positive effect on the accuracy of the model. We can see that the Mean MSE was 157.2 with 50 epochs (Part B) and Mean MSE with 100 epochs is 64.4! And the standard deviation of MSE with 100 epochs is almost 6 times smaller than with 50 epochs.

### Implementing 3 hidden layers instead of 1 also has increased the accuracy of the model greatly, almost as good as in Part C where we doubled the number of epochs. 